src/twisted/web/test/test_agent.py

```diff
@@ -3216,6 +3216,24 @@ def test_cancel(self):
         self.failureResultOf(deferred, defer.CancelledError)
         self.assertTrue(response.transport.aborting)
 
+    def test_cancelWithConnectionLost(self):
+        """
+        When cancelling the L{Deferred} returned by L{client.readBody}, the
+        CancelledError failure is sent repeatedly in response protocol. Exception
+        defer.AlreadyCalledError should be suppressed
+        """
+        response = DummyResponse()
+        deferred = client.readBody(response)
+        deferred.cancel()
+
+        try:
+            response.protocol.connectionLost(Failure(defer.CancelledError))
```

Comment:
I think that we should not use CancelledError for the connection lost.... we usually have the connection closed with ConnectionDone or other connection related errors.


Note that we also need a separate test for `PartialDownloadError` which can happen if there is a TCP timeout 

```suggestion
        # At some point the connection used by the response
        # to read the body will get lost.
        # This happend after the `readBody` deferred is
        # cancelled by the user.
        response.protocol.connectionLost(Failure(Exception('We are lost'))
```

Reply:
1. Yes. CancelledError doesn't fit here. It can be ConnectionDone (in case if connection was properly closed) or ConnectionLost (if connection was lost in a non-clean fashion) or possibly something else.
2. When you say about separate test for PartialDownloadError you mean the same test as test_withPotentialDataLoss but with cancel?

Reply:
@adiroiban I used your version for exception type